python/python/lancedb/table.py

```diff
@@ -139,30 +142,26 @@ def _into_pyarrow_table(data) -> pa.Table:
         )
 
 
-def _iterator_to_table(data: Iterable) -> pa.Table:
-    batches = []
-    schema = None  # Will get schema from first batch
-    for batch in data:
-        batch_table = _into_pyarrow_table(batch)
-        if schema is not None:
-            if batch_table.schema != schema:
+def _iterator_to_table(data: Iterable) -> pa.RecordBatchReader:
+    first = _into_pyarrow_reader(next(data))
+    schema = first.schema
+
+    def gen():
+        yield from first
+        for batch in data:
+            table = _into_pyarrow_reader(batch).read_all()
+            if table.schema != schema:
                 try:
-                    batch_table = batch_table.cast(schema)
+                    table = table.cast(schema)
```

Comment:
Is `cast` a valid thing to call on a reader?

Reply:
We called `read_all()` to make it a table.